tests/integration_tests/cmd/test_status.py

```diff
@@ -82,3 +82,57 @@ def test_status_json_errors(client):
     assert "Invalid cloud-config provided" in json.loads(status_json)[
         "recoverable_errors"
     ].get("WARNING").pop(0)
+
+
+EARLY_BOOT_WAIT_USER_DATA = """\
+#cloud-config
+runcmd: [systemctl enable before-cloud-init-local.service]
+write_files:
+- path: /waitoncloudinit.sh
+  permissions: '0755'
+  content: |
+    #!/bin/sh
+    [ -f /var/lib/cloud/data/status.json ] && touch /$1.start-hasstatusjson || touch /$1.start-nostatusjson
+    cloud-init status --wait --long > $1
+- path: /lib/systemd/system/before-cloud-init-local.service
+  permissions: '0644'
+  content: |
+    [Unit]
+    Description=BEFORE cloud-init local
+    DefaultDependencies=no
+    Before=cloud-init-local.service
+    Before=shutdown.target
+    Before=sysinit.target
+    Conflicts=shutdown.target
+    RequiresMountsFor=/var/lib/cloud
+
+    [Service]
+    Type=simple
+    ExecStart=/waitoncloudinit.sh /before-local
+    RemainAfterExit=yes
+    TimeoutSec=0
+
+    # Output needs to appear in instance console output
+    StandardOutput=journal+console
+
+    [Install]
+    WantedBy=cloud-init.target
+"""  # noqa: E501
+
+
+@pytest.mark.user_data(EARLY_BOOT_WAIT_USER_DATA)
+@pytest.mark.lxd_use_exec
+def test_status_block_through_all_boot_status(client):
+    """Assert early boot cloud-init status --wait does not exit early."""
+    client.execute("cloud-init clean --logs --reboot")
+    wait_for_cloud_init(client).stdout.strip()
+    status_json = client.execute("cloud-init status --wait --format=json")
+
+    # Assert that before-cloud-init-local.service started before
+    # cloud-init-local.service could create status.json
+    assert client.execute("test -f /before-local.start-nostatusjson")
+    # Assert the file created at the end of
+    # before-cloud-init-local.service is newer than the last log entry in
+    # /var/log/cloud-init.log
+    assert client.execute("test /before-local -nt /var/log/cloud-init.log").ok
```

Comment:
Depending on timestamps could be unreliable since reads may update timestamps for any filesystems mounted without `noatime` or `A` attribute set, since any external tool that reads `/var/log/cloud-init.log` would update the timestamp of the file. Maybe the file could contain a time for a more robust test?

Reply:
Good suggestion, I've adapted the unit to echo an ending timestamp and validate that the time is later than the timestamp of the last log exposed by `cloud-init analyze dump`.